API/repository/base.go

```diff
@@ -0,0 +1,193 @@
+package repository
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	u "p3/utils"
+	"time"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/mongo/readpref"
+)
+
+// Database
+var globalDB *mongo.Database
+var globalClient *mongo.Client
+
+func ConnectToDB(host, port, user, pass, dbName string) error {
+	client, err := ConnectToMongo(host, port, user, pass, dbName)
+	if err != nil {
+		return err
+	}
+
+	globalClient = client
+
+	db, err := GetDatabase(client, dbName)
+	if err != nil {
+		return err
+	}
+
+	globalDB = db
+
+	err = SetupDB(db)
+	if err != nil {
+		return err
+	}
+
+	err = createInitialData(db, dbName)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func SetupDB(db *mongo.Database) error {
+	// Indexes creation
+	// Enforce unique children
+	for _, entity := range []int{u.DOMAIN, u.SITE, u.BLDG, u.ROOM, u.RACK, u.DEVICE, u.AC, u.PWRPNL, u.CABINET, u.CORRIDOR, u.GROUP, u.STRAYOBJ} {
+		if err := createUniqueIndex(db, entity, bson.M{"id": 1}); err != nil {
```

Comment:
Why can it be here? As you say, it will be executed every time the api is launched, but if the index already exists it doesn't generate any problem, it just ignores the creation.
Why should it be here? Because the index creation is not an operational task of the database independent of the api operation (as it is for example to create the first user). If the indexes do not exist the api does not work correctly, because the only verification that the ids are not repeated is thanks to these indexes. Therefore, the api must make sure that the indices exist to make sure it runs on a base that has the right characteristics to work properly.
Why do I need it to be here? As I am adding integration tests, I need to clean the database between runs. To do this, I drop the database and then reconfigure it, which is much faster than deleting all the elements of a database.

Reply:
Both approaches seem ok to me, because createdb.js is still mandatory in any case, it should always be run before the API and just once, so the idea of creating collections and indexes there (which should always be done only once) makes sense (the API wouldn't work properly in any case without createdb.js first). But the tests do make it a stronger argument.